format-avro/src/main/java/io/cdap/plugin/format/avro/AvroToStructuredTransformer.java

```diff
@@ -91,9 +98,96 @@ public Schema convertSchema(org.apache.avro.Schema schema) throws IOException {
     if (schemaCache.containsKey(hashCode)) {
       structuredSchema = schemaCache.get(hashCode);
     } else {
-      structuredSchema = Schema.parseJson(schema.toString());
+      String strSchema = schema.toString();
+      String jsonSchema = preprocessSchema(GSON.fromJson(strSchema, JsonObject.class)).toString();
+      structuredSchema = Schema.parseJson(jsonSchema);
       schemaCache.put(hashCode, structuredSchema);
     }
     return structuredSchema;
   }
+
+  private JsonObject preprocessSchema(JsonObject schema) {
+
+    if (!schema.has("type")) {
+      return schema;
+    }
+
+    JsonElement type = schema.get("type");
+
+    if (type.isJsonArray()) {   // Union
+      JsonArray processedUnion = preprocessUnion(type.getAsJsonArray());
+      schema.remove("type");
+      schema.add("type", processedUnion);
+    } else if (type.isJsonObject()) {   // Unnamed Complex type
+
+      preprocessSchema(type.getAsJsonObject());
+    } else {
+      String typeName = type.getAsString();
+
+      switch (typeName) {
+        case "record":
+          for (JsonElement field : schema.get("fields").getAsJsonArray()) {
+            preprocessSchema(field.getAsJsonObject());
+          }
+          customTypes.put(schema.getAsJsonPrimitive("name").getAsString(), schema);
+          break;
+
+        case "map":
+          schema.addProperty("keys", "string");
+          JsonElement value = schema.get("values");
+          processElementInSchema(schema, value, "values");
+          break;
+
+        case "array":
+          JsonElement items = schema.get("items");
+          processElementInSchema(schema, items, "items");
+          break;
+
+        case "enum":
+          customTypes.put(schema.getAsJsonPrimitive("name").getAsString(), schema);
+          break;
+
+        default:
+          if (customTypes.containsKey(typeName)) {
+            schema.remove("type");
+            schema.add("type", customTypes.get(typeName));
+          }
```

Comment:
Are you dropping other things like primitive types? E.g. how somelhing like `"fields" : [
	      {"name": "a", "type": "long"},
	      {"name": "b", "type": "string"}
	      ]` work?